src/Decompiler/Structure/FictitiousGlobalEliminator.cs

```diff
@@ -0,0 +1,66 @@
+#region License
+/* 
+ * Copyright (C) 1999-2020 Pavel Tomin.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
+#endregion
+
+using Reko.Core;
+using Reko.Core.Code;
+using Reko.Core.Expressions;
+using Reko.Core.Types;
+
+namespace Reko.Structure
+{
+    /// <summary>
+    /// Eliminate fictitious "global" identifiers. For instance, transform
+    /// "global->ptr0001" to "g_ptr0001" or "global->userVariable" to
+    /// "userVariable"
+    /// </summary>
+    public class FictitiousGlobalEliminator : InstructionTransformer
+    {
+        private readonly Program program;
+        private readonly Procedure proc;
+
+        public FictitiousGlobalEliminator(Program program, Procedure proc)
+        {
+            this.program = program;
+            this.proc = proc;
+        }
+
+        public void Transform()
+        {
+            foreach (var stm in proc.Statements)
+            {
+                stm.Instruction = Transform(stm.Instruction);
+            }
+        }
+
+        public override Expression VisitFieldAccess(FieldAccess acc)
```

Comment:
The logic is fine, but do you agree that it makes sense to add this logic to `ComplexExpressionBuilder` and `TypedConstantRewriter` instead? That would avoid injecting references to the temporary `global` in the first place.

Reply:
Done in 8ae7402